File: test_quick.c
#include "fht.h"
#include <stdio.h>

static int test_defines(void) {
    printf("%s\n", __func__);

    #ifdef __AVX__
    printf("__AVX__ is defined\n");
    #endif

    #ifdef __AVX2__
    printf("__AVX2__ is defined\n");
    #endif

    #ifdef __SSE2__
    printf("__SSE2__ is defined\n");
    #endif

    #ifdef __aarch64__
    printf("__aarch64__ is defined\n");
    #endif

    #ifdef __ARM_NEON
    printf("__ARM_NEON is defined\n");
    #endif
    return 0;
}

static int test_fast_copy(void) {
    printf("\n%s\n", __func__);

    float input[4] = {1.0, 2.0, 3.0, 4.0};
    float output[4] = {0.0, 0.0, 0.0, 0.0};

    printf("Input:  [%f, %f, %f, %f]\n", input[0], input[1], input[2], input[3]);
    printf("Output before: [%f, %f, %f, %f]\n", output[0], output[1], output[2], output[3]);

    fast_copy(output, input, 4 * sizeof(float));

    printf("Output after:  [%f, %f, %f, %f]\n", output[0], output[1], output[2], output[3]);

    return 0;
}

static int test_inplace(void) {
    printf("\n%s\n", __func__);

    float data[4] = {1.0, -1.0, 1.0, -1.0};

    printf("Input:  [%f, %f, %f, %f]\n", data[0], data[1], data[2], data[3]);

    int result = fht_float(data, 2);

    printf("Output: [%f, %f, %f, %f]\n", data[0], data[1], data[2], data[3]);
    printf("Return value: %d\n", result);

    return 0;
}

static int test_oop(void) {
    printf("\n%s\n", __func__);

    float input[4] = {1.0, -1.0, 1.0, -1.0};
    float output[4] = {0.0, 0.0, 0.0, 0.0};

    printf("Input:  [%f, %f, %f, %f]\n", input[0], input[1], input[2], input[3]);
    printf("Output before: [%f, %f, %f, %f]\n", output[0], output[1], output[2], output[3]);

    int result = fht_float_oop(input, output, 2);

    printf("Output after:  [%f, %f, %f, %f]\n", output[0], output[1], output[2], output[3]);
    printf("Return value: %d\n", result);

    return 0;
}

int main(void) {
    test_defines();
    test_fast_copy();
    test_inplace();
    test_oop();
    return 0;
}

File: test_quick.py
#!/usr/bin/env python3
"""
Python test corresponding to test_quick.c
Tests the FFHT Python wrapper to verify C implementation is correct
"""

import numpy as np
import ffht

def test_inplace():
    """Test in-place FHT (corresponds to test_inplace() in test_quick.c)"""
    print("\ntest_inplace")

    # Same test data as C: {1.0, -1.0, 1.0, -1.0}
    data = np.array([1.0, -1.0, 1.0, -1.0], dtype=np.float32)

    print(f"Input:  {data}")

    # In-place transform (modifies data)
    ffht.fht(data)

    print(f"Output: {data}")

    # Verify the results
    # For input [1, -1, 1, -1] with log_n=2, what should the output be?
    # Let's see what the C test outputs
    return data

def test_inplace_copy():
    """
    Test that mimics out-of-place by copying first
    (Python wrapper only has in-place, so we simulate OOP)
    """
    print("\ntest_inplace_copy (simulating OOP)")

    input_data = np.array([1.0, -1.0, 1.0, -1.0], dtype=np.float32)
    output = input_data.copy()  # Create a copy

    print(f"Input:  {input_data}")
    print(f"Output before: {output}")

    # Transform the copy (simulates out-of-place)
    ffht.fht(output)

    print(f"Output after:  {output}")

    # Verify input unchanged
    assert np.array_equal(input_data, [1.0, -1.0, 1.0, -1.0])

    return output

def test_double():
    """Test with float64 (double precision)"""
    print("\ntest_double")

    data = np.array([1.0, -1.0, 1.0, -1.0], dtype=np.float64)

    print(f"Input:  {data}")

    ffht.fht(data)

    print(f"Output: {data}")

    return data

def test_larger_size():
    """Test with size 8"""
    print("\ntest_larger_size")

    data = np.array([1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0], dtype=np.float32)

    print(f"Input:  {data}")

    ffht.fht(data)

    print(f"Output: {data}")

    return data

def main():
    print("=" * 60)
    print("FFHT Python Test (corresponding to test_quick.c)")
    print("=" * 60)

    result1 = test_inplace()
    result2 = test_inplace_copy()
    result3 = test_double()
    result4 = test_larger_size()

    print("\n" + "=" * 60)
    print("Summary:")
    print("=" * 60)
    print(f"test_inplace result:      {result1}")
    print(f"test_inplace_copy result: {result2}")
    print(f"test_double result:       {result3}")
    print(f"test_larger_size result:  {result4}")

    # Check that float32 and float64 give same results (within precision)
    print("\nVerifying float32 vs float64 consistency:")
    if np.allclose(result1, result3, rtol=1e-6):
        print("✓ Results match within tolerance")
    else:
        print("✗ Results differ!")
        print(f"  Difference: {result1 - result3}")

if __name__ == "__main__":
    main()
